fdbserver/SimEncryptVaultProxy.actor.cpp

```diff
@@ -0,0 +1,153 @@
+#include "flow/ActorCollection.h"
+#include "flow/Error.h"
+#include "flow/IRandom.h"
+#include "flow/ITrace.h"
+#include "flow/StreamCipher.h"
+#include "flow/UnitTest.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+struct SimEncryptKeyCtx {
+	SimEncryptKeyId id;
+	SimEncryptKey key;
+
+	SimEncryptKeyCtx() : id(0) {}
+	explicit SimEncryptKeyCtx(SimEncryptKeyId kId, const char* data) : id(kId), key(data) {}
+};
+
+struct SimEncyrptVaultProxyContext {
+	uint32_t maxEncryptionKeys;
+	std::unordered_map<SimEncryptKeyId, std::unique_ptr<SimEncryptKeyCtx>> simEncryptKeyStore;
+
+	SimEncyrptVaultProxyContext() : maxEncryptionKeys(0) {}
+	explicit SimEncyrptVaultProxyContext(uint32_t keyCount) : maxEncryptionKeys(keyCount) {
+		uint8_t buffer[AES_256_KEY_LENGTH];
+
+		// Construct encryption keyStore.
+		for (int i = 0; i < maxEncryptionKeys; i++) {
+			generateRandomData(&buffer[0], AES_256_KEY_LENGTH);
+			SimEncryptKeyCtx ctx(i, reinterpret_cast<const char*>(buffer));
+			simEncryptKeyStore[i] = std::make_unique<SimEncryptKeyCtx>(i, reinterpret_cast<const char*>(buffer));
+		}
+	}
+};
+
+ACTOR Future<Void> simEncryptVaultProxyCore(SimEncryptVaultProxyInterface interf, uint32_t maxEncryptKeys) {
+	state SimEncyrptVaultProxyContext vaultProxyCtx(maxEncryptKeys);
+
+	ASSERT(vaultProxyCtx.simEncryptKeyStore.size() == maxEncryptKeys);
+
+	TraceEvent("SimEncryptVaultProxy_Init", interf.id()).detail("MaxEncrptKeys", maxEncryptKeys);
+
+	loop {
+		choose {
+			when(SimGetEncryptKeyByKeyIdRequest req = waitNext(interf.encryptKeyLookupByKeyId.getFuture())) {
+				SimGetEncryptKeyByKeyIdReply reply;
+
+				// Lookup corresponding EncryptKeyCtx for input keyId
+				if (vaultProxyCtx.simEncryptKeyStore.find(req.encryptKeyId) != vaultProxyCtx.simEncryptKeyStore.end()) {
+					reply.encryptKey = StringRef(vaultProxyCtx.simEncryptKeyStore[req.encryptKeyId].get()->key);
+					req.reply.send(reply);
+				} else {
+					req.reply.sendError(key_not_found());
+				}
+			}
+			when(SimGetEncryptKeyByDomainIdRequest req = waitNext(interf.encryptKeyLookupByDomainId.getFuture())) {
+				SimGetEncryptKeyByDomainIdReply reply;
+
+				// Map encryptionDomainId to corresponding EncryptKeyCtx element using a modulo operation. This would
+				// mean multiple domains gets mapped to the same encryption key which is fine, the EncryptKeyStore
+				// guarantees that keyId -> plaintext encryptKey mapping is idempotent.
+
+				reply.encryptKeyId = req.encryptDomainId % maxEncryptKeys;
+				reply.encryptKey = StringRef(vaultProxyCtx.simEncryptKeyStore[reply.encryptKeyId].get()->key);
+				req.reply.send(reply);
+			}
+		}
+	}
+}
+
+void forceLinkSimEncryptVaultProxyTests() {}
+
+namespace {
+
+ACTOR Future<Void> testRunWorkload(SimEncryptVaultProxyInterface inf, uint32_t nEncryptionKeys) {
+	state uint32_t maxEncryptionKeys = nEncryptionKeys;
+	state int maxDomainIds = deterministicRandom()->randomInt(121, 295);
```

Comment:
These numbers seem like magic numbers, what is their significance, if any?

Reply:
As you mentioned they are magic numbers, test is where we can innovative about picking numbers just out of a hat !:)